Use more WebRender types in gfx/display_list.

pyfisch opened the pull request:
Use more WebRender types in gfx/display_list.
This uses floating-point (Layout) coordinates in where possible.
Replace NormalBorder struct with WebRender equivalent.
Remove ToPointF and ToRectF traits.
Convert border RepeatKeyword with ToLayout.
Add some definitions to malloc_size_of for WebRender types.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/19782)
<!-- Reviewable:end -->

pyfisch committed:
Use more WebRender types in gfx/display_list

This uses floating-point (Layout) coordinates in where possible.
Replace NormalBorder struct with WebRender equivalent.
Remove ToPointF and ToRectF traits.
Convert border RepeatKeyword with ToLayout.
Add some definitions to malloc_size_of for WebRender types.

highfive commented:
Heads up! This PR modifies the following files:
 * @emilio: components/layout/display_list/webrender_helpers.rs, components/layout/display_list/builder.rs, components/layout/display_list/conversions.rs, components/layout/display_list/background.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify gfx and layout code, but no tests are modified. Please consider adding a test!

mrobinson commented on components/gfx/display_list/mod.rs:
```diff
@@ -423,7 +423,11 @@ impl BaseDisplayItem {
             },
-            local_clip: LocalClip::from(max_rect().to_rectf()),
+            // Create a rectangle of maximal size.
+            local_clip: LocalClip::from(LayoutRect::new(
```
It probably makes sense to implement max_rect for LayoutRect if we can. Perhaps this can be a trait.

pyfisch replied:
There is a trait [`MaxRect`](https://doc.servo.org/webrender/util/trait.MaxRect.html) in webrender but I can't use it from servo.

Maybe this trait should instead live in euclid?

mrobinson replied:
I think it makes sense to at least copy the trait from WebRender and replace max_rect with it.

mrobinson commented on components/gfx/display_list/mod.rs:
```diff
@@ -680,15 +684,15 @@ pub struct ImageDisplayItem {
     /// The algorithm we should use to stretch the image. See `image_rendering` in CSS-IMAGES-3 §
     /// 5.3.
-    pub image_rendering: image_rendering::T,
+    pub image_rendering: webrender_api::ImageRendering,
```
We are importing the other webrender_api types directly in this file, so we should probably be consistent here and import ImageRendering directly as well.

mrobinson commented on components/gfx/display_list/mod.rs:
```diff
@@ -701,16 +705,16 @@ pub struct IframeDisplayItem {
-    /// True if gradient repeats infinitly.
-    pub repeating: bool,
+    /// Whether the gradient is repeated or clamped.
+    pub extend: webrender_api::ExtendMode,
```
Same comment here and other places in the file as well. I would also call this `extend_mode` in order to make the code more readable.

mrobinson commented on components/layout/display_list/background.rs:
```diff
@@ -370,6 +370,14 @@ fn convert_gradient_stops(gradient_items: &[GradientItem], total_length: Au) ->
     stops
 }
 
+fn as_gradient_extend(repeating: bool) -> ExtendMode {
```
Probably better to call this as `as_gradient_extend_mode` since, in my opinion, it makes it more obvious what this function does.

mrobinson commented on components/layout/display_list/builder.rs:
```diff
@@ -826,6 +816,20 @@ fn calculate_inner_bounds(mut bounds: Rect<Au>, offsets: SideOffsets2D<Au>) -> R
+fn simple_normal_border(
+    color: ColorF,
+    style: webrender_api::BorderStyle,
+) -> webrender_api::NormalBorder {
```
Again, I'd say just import NormalBorder directly.

mrobinson commented on components/layout/display_list/builder.rs:
```diff
@@ -1374,20 +1378,31 @@ impl FragmentDisplayListBuilding for Fragment {
             display_list_section,
         );
 
+        let radius_tmp = build_border_radius(&bounds, border_style_struct);
```
Why not just call this `radius` or `border_radius`?

mrobinson commented on components/gfx/display_list/mod.rs:
```diff
@@ -728,24 +732,24 @@ pub struct GradientDisplayItem {
-    /// True if gradient repeats infinitly.
-    pub repeating: bool,
+    /// Whether the gradient is repeated or clamped.
+    pub extend: webrender_api::ExtendMode,
```
I would cal this `extend_mode` in order to make the code more readable.

mrobinson reviewed (changes requested):
This is great work! I'm really glad to see us moving iteratively toward building a WebRender display list directly. In general, this looks very good to me, but I have a few comments on some minor details that would be good to change.

mrobinson commented:
I would also go ahead and title this commit 'Use more WebRender types in gfx/display_list' since I think that's a really excellent summary of what it does.

pyfisch commented:
All comments addressed except for `max_rect`.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #19780) made this pull request unmergeable. Please resolve the merge conflicts.

pyfisch committed:
Introduce MaxRect trait

It is implemented for LayoutRect and Rect<Au>.
Replaces the max_rect() function from servo_geometry.

pyfisch commented:
Rebased and updated. Copied the MaxRect trait.

mrobinson reviewed (approved):
This looks great! Thank you for the cleanups.

mrobinson commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit af52233 has been approved by `mrobinson`

<!-- @bors-servo r=mrobinson af52233ae57af8c05eab1f60f208cf5b42a832bd -->

bors-servo commented:
:hourglass: Testing commit af52233ae57af8c05eab1f60f208cf5b42a832bd with merge aee0d694cc077d4694fc3114af3a2ad83e326cbf...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10433), [arm32](http://build.servo.org/builders/arm32/builds/10670), [arm64](http://build.servo.org/builders/arm64/builds/10696), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10722), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7585), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7598), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10802), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3206), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3226), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/7475), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7537), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3296), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3245), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6099)
Approved by: mrobinson
Pushing aee0d694cc077d4694fc3114af3a2ad83e326cbf to master...
